### PR TITLE
[WIP] Add partial support for FreeBSD systems

### PR DESCRIPTION
**Pull request by Rogiel**
Known issues:

CFPlatform.c
=======
 - `_CFThreadSetName` and `_CFThreadGetName`: use a thread local? pthread specific?
 - `_CFEnviron`: environ linkage fails, even though the documentation states it should be available.

CFRunLoop.c
=======
 - Only stubs, not working yet.

CFUtilities.c
=======
 - `__CFGetEnvironment`: same as _CFEnviron

### CODE REVIEW

**Comment by Rogiel** [comment]
Would like to get some early reviews and a call for help if someone else is interested in getting Foundation working under FreeBSD.

File: CoreFoundation/Base.subproj/CFBase.h [hunk 1]
```diff
@@ -70,6 +70,10 @@
 
 #if __BLOCKS__ && ((TARGET_OS_MAC && !(TARGET_OS_EMBEDDED || TARGET_OS_IPHONE)) || (TARGET_OS_EMBEDDED || TARGET_OS_IPHONE))
 #include <Block.h>
+#endif
+
+#if defined(__FreeBSD__)
```
Comment (millenomi) [inline CoreFoundation/Base.subproj/CFBase.h]: This should really be DEPLOYMENT_TARGET_FREEBSD (and then that should be defined in terms of __FreeBSD__.)

File: CoreFoundation/Base.subproj/CFInternal.h [hunk 2]
```diff
@@ -596,7 +596,11 @@ static _Bool __os_warn_unused(_Bool x) { return x; }
 
 #if !__HAS_DISPATCH__
 
+#if defined(__FreeBSD__)
+typedef long dispatch_once_t;
```
Comment (millenomi) [inline CoreFoundation/Base.subproj/CFInternal.h]: What is the rationale behind this?

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 3]
```diff
@@ -155,16 +155,26 @@ const char *_CFProcessPath(void) {
...
 
+#if DEPLOYMENT_TARGET_FREEBSD
+#include <pthread_np.h>             // for _CFIsMainThread
+#endif
+
 Boolean _CFIsMainThread(void) {
+#if DEPLOYMENT_TARGET_LINUX
     return syscall(SYS_gettid) == getpid();
+#elif DEPLOYMENT_TARGET_FREEBSD
+	return pthread_main_np() != 0;
```
Comment (millenomi) [inline CoreFoundation/Base.subproj/CFPlatform.c]: There are whitespace mismatches — make sure your editor matches the settings of the file around it?

File: CoreFoundation/Base.subproj/CFPriv.h [hunk 4]
```diff
@@ -575,6 +575,12 @@ CF_INLINE CFAbsoluteTime _CFAbsoluteTimeFromFileTimeSpec(struct timespec ts) {
     return (CFAbsoluteTime)((CFTimeInterval)ts.tv_sec - kCFAbsoluteTimeIntervalSince1970) + (1.0e-9 * (CFTimeInterval)ts.tv_nsec);
 }
 
+// For whatever reason, including <time.h> is not defining time_t on FreeBSD
```
Comment (millenomi) [inline CoreFoundation/Base.subproj/CFPriv.h]: I see it defined in https://github.com/freebsd/freebsd/blob/master/include/time.h — can you investigate it?

File: CoreFoundation/Base.subproj/CFUtilities.c [hunk 5]
```diff
@@ -1494,14 +1494,19 @@ CF_EXPORT Boolean _CFExtensionIsValidToAppend(CFStringRef extension) {
...
+#else
 CFDictionaryRef __CFGetEnvironment() {
     static dispatch_once_t once = 0L;
     static CFMutableDictionaryRef envDict = NULL;
     dispatch_once(&once, ^{
 #if DEPLOYMENT_TARGET_MACOSX || DEPLOYMENT_TARGET_EMBEDDED
         extern char ***_NSGetEnviron(void);
         char **envp = *_NSGetEnviron();
-#elif DEPLOYMENT_TARGET_FREEBSD || TARGET_OS_CYGWIN
+#elif TARGET_OS_CYGWIN || DEPLOYMENT_TARGET_FREEBSD
```
Comment (millenomi) [inline CoreFoundation/Base.subproj/CFUtilities.c]: ?

File: CoreFoundation/Base.subproj/CFBase.h [hunk 6]
```diff
@@ -70,6 +70,10 @@
...
+#include <libutil.h>
```
Comment (parkera) [inline CoreFoundation/Base.subproj/CFBase.h]: Why would we need to include this in a key public header?

File: CoreFoundation/Base.subproj/CFInternal.h [hunk 7]
```diff
@@ -596,7 +596,11 @@ static _Bool __os_warn_unused(_Bool x) { return x; }
 
 #if !__HAS_DISPATCH__
 
+#if defined(__FreeBSD__)
```
Comment (parkera) [inline CoreFoundation/Base.subproj/CFInternal.h]: For the internal headers we should `DEPLOYMENT_TARGET` style macros.

File: CoreFoundation/Base.subproj/ForFoundationOnly.h [hunk 8]
```diff
@@ -13,6 +13,13 @@
 #if !defined(__COREFOUNDATION_FORFOUNDATIONONLY__)
 #define __COREFOUNDATION_FORFOUNDATIONONLY__ 1
 
+/// FreeBSD requires this declaration on the top. Some other include below this
```
Comment (millenomi) [inline CoreFoundation/Base.subproj/ForFoundationOnly.h]: Can you also investigate here? In general, I don't think it's wise to mess with system imports in the CF headers, or to redefine system types, until and unless we are 100% sure that what's happening is not avoidable.

File: CoreFoundation/Base.subproj/CFPlatform.c [hunk 9]
```diff
@@ -1327,6 +1337,11 @@ CF_SWIFT_EXPORT int _CFThreadSetName(pthread_t thread, const char *_Nonnull name
     return EINVAL;
 #elif DEPLOYMENT_TARGET_LINUX
     return pthread_setname_np(thread, name);
+#elif DEPLOYMENT_TARGET_FREEBSD
+	pthread_set_name_np(thread, name);
+	return 0;
+#else
```
Comment (parkera) [inline CoreFoundation/Base.subproj/CFPlatform.c]: Nit pick, but it's best to try and preserve indentation as-is in these files.

File: CoreFoundation/Base.subproj/CFUtilities.c [hunk 10]
```diff
@@ -1494,14 +1494,19 @@ CF_EXPORT Boolean _CFExtensionIsValidToAppend(CFStringRef extension) {
 
 #if DEPLOYMENT_RUNTIME_SWIFT
 
+#if DEPLOYMENT_TARGET_FREEBSD
```
Comment (parkera) [inline CoreFoundation/Base.subproj/CFUtilities.c]: I know you listed as a known issue, but we'll have to resolve this before merging. It's used in quite a few places.

**Comment by millenomi** [comment]
@Rogiel this has seen no change, so I'm closing it. If you want to upstream a patch for this again, ping me?

**Pull request closed without merging**